### ISSUE TITLE
Build fails on OS X with undefined 'contours' error

### ISSUE THREAD

**Issue by bitserf** [be5invis/Iosevka#15]
Hi,

I'm trying to build the TTF files on OS X.

I installed the dependencies:

```shell
$ brew list
autoconf		harfbuzz		pango
automake		icu4c			pinentry
brew-cask		jpeg			pixman
cairo			kafka			pkg-config
carthage		kafkacat		postgresql
czmq			libassuan		pth
dirmngr			libffi			python
fish			libgcrypt		readline
fontconfig		libgpg-error		sbt
*fontforge*		libksba			scala
fonttools		libpng			sqlite
freetype		librdkafka		tree
gdbm			lib tiff			*ttfautohint*
gettext			libtool			typesafe-activator
glib			libusb			xz
gnupg2			libusb-compat		zeromq
gobject-introspection	lzlib			zookeeper
gpg-agent		*node*
cradle			openssl
```

`ttx` is also executable from the command-line.

Then installed patel:

```shell
$ npm install patel -g
patel@0.19.2 /usr/local/lib/node_modules/patel
├── patrisika@0.19.2 (patrisika-scopes@0.9.2)
├── resumer@0.0.0 (through@2.3.8)
├── escodegen@1.7.0 (estraverse@1.9.3, esutils@2.0.2, esprima@1.2.5, optionator@0.5.0, source-map@0.2.0)
└── yargs@3.24.0 (decamelize@1.0.0, camelcase@1.2.1, window-size@0.1.2, y18n@3.1.0, os-locale@1.4.0, cliui@2.1.0)
```

Then installed the dependencies:

```shell
$ npm install
libspiro-js@0.2.1 node_modules/libspiro-js

bezier-js@1.0.8 node_modules/bezier-js

toml@2.3.0 node_modules/toml

yargs@3.24.0 node_modules/yargs
├── decamelize@1.0.0
├── camelcase@1.2.1
├── window-size@0.1.2
├── y18n@3.1.0
├── os-locale@1.4.0 (lcid@1.0.0)
└── cliui@2.1.0 (wordwrap@0.0.2, right-align@0.1.3, center-align@0.1.1)

node-sfnt@0.0.20 node_modules/node-sfnt
└── xmldom@0.1.19
```

And ran make: 

```shell
$ make
patel-c --strict support/glyph.patel -o support/glyph.js
patel-c --strict support/stroke.patel -o support/stroke.js
patel-c --strict support/spiroexpand.patel -o support/spiroexpand.js
patel-c --strict parameters.patel -o parameters.js
patel-c --strict buildglyphs.patel -o buildglyphs.js
node generate regular build/.pass0-iosevka-regular.ttf --dumpmap build/iosevka-regular.charmap --dumpfeature build/.pass0-iosevka-regular.ab.fea
/Users/ljb/Code/Iosevka/support/glyph.js:223
        r19_contours = r19_glyph.contours;
                                ^
TypeError: Cannot read property 'contours' of undefined
    at _r0_t3._r0_t16 (/Users/ljb/Code/Iosevka/support/glyph.js:223:33)
    at _r0_t3._r4753_t3 (/Users/ljb/Code/Iosevka/buildglyphs.js:11942:40)
    at _r9_t44 (/Users/ljb/Code/Iosevka/buildglyphs.js:356:29)
    at _r9_t784 (/Users/ljb/Code/Iosevka/buildglyphs.js:11926:20)
    at Object._r0_t4 [as build] (/Users/ljb/Code/Iosevka/buildglyphs.js:11947:9)
    at Object.<anonymous> (/Users/ljb/Code/Iosevka/generate.js:25:27)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
make: *** [build/.pass0-iosevka-regular.ttf] Error 1
```

Any suggestions?

My OS X version is 10.11 (El Capitan), but it also failed on 10.10 (Yosemite).

Thanks!

**Comment by be5invis** [comment]
It's during building ligatures. But this should not happen.
What's the version of Node you have. I'm using 4.0.0 and everything works OK.
Could you please uncomment line 255 of `buildglyphs.patel`, remove the leading `#` and show the last glyph being built?

**Comment by bitserf** [comment]
No problem - There wasn't a Node 4.0 on OS X Homebrew package manager yet so I upgraded to it from source, but I still get the error - Is one of my dependencies wrong version maybe?

After uncommenting the line in buildglyphs.js, here is the partial output:

```shell
Building /CoverU for Iosevka Regular
Building /numC2 for Iosevka Regular
Building /denU for Iosevka Regular
Building /LoverN for Iosevka Regular
Building /numL for Iosevka Regular
Building /denN for Iosevka Regular
Building /spaceoverspace for Iosevka Regular
Building /numspace for Iosevka Regular
Building /denspace for Iosevka Regular
Building /D_Zcaron for Iosevka Regular
/Users/ljb/Code/Iosevka/support/glyph.js:223
        r19_contours = r19_glyph.contours;
```

The buildglyphs.js is in this gist: https://gist.github.com/bitserf/2adfe27b39e141bc88fe

**Comment by be5invis** [comment]
It's weird -- your file works correctly on my computer.

**Comment by be5invis** [comment]
![](http://ww4.sinaimg.cn/large/798f7769gw1ew0okf734kj21f411iqs3.jpg)

I asked one of my friends to build this and it is completely OK. He installed same dependencies and successfully built `build/.pass0-iosevka-regular.ttf`, which is the file your build fails.

**Comment by bitserf** [comment]
I am in an English locale:

```shell
LANG=en_NZ.UTF-8
LC_CTYPE=en_NZ.UTF-8
```

Despite not knowing what I'm doing in .patel files, I added some logging statements to `createLigature` of `autobuilds.patel`, and it looks like this now:

```
define [createLigature unicode] : begin {
        local ids : ().slice.call arguments 1
        local glyphName : ids.join '_'
        process.stderr.write : "glyphName is /" + glyphName + "\n"
        if glyphs.(glyphName) : begin {
                local j 2
                while glyphs.[glyphName + j] : inc j
                set glyphName [glyphName + j]
                process.stderr.write : "glyphName is now /" + glyphName + "\n"
        }
        create-glyph glyphName : glyph-construction {
                process.stderr.write : "unicode: " + unicode + "\n"
                if unicode : assign-unicode unicode
                set-width : ids.length * WIDTH
                foreach id [items-of : ids.reverse] : begin {
                        process.stderr.write : "id: " + id + "\n"
                        apply-transform : Translate WIDTH 0
                        process.stderr.write : "transform translate applied: " + id + "\n"
                        include glyphs.(id)
                        process.stderr.write : "glyphs included: " + id + "\n"
                }
                process.stderr.write : "glyph: " + glyphName + " created\n"
        }
}
```

Here is the output when it fails:

```shell
glyphName is /D_Zcaron
Building /D_Zcaron for Iosevka Regular
unicode: 452
id: Zcaron
transform translate applied: Zcaron
/Users/ljb/Code/Iosevka/support/glyph.js:223
        r19_contours = r19_glyph.contours;
```

As you can see it is the `include glyphs.(id)` that fails, if I comment it out, then the build is successful.

I really want to figure out what's going wrong now though, because I am interested in both building Iosevka and ligatures :)

Any further hints?

**Comment by bitserf** [comment]
I dumped the contents of the 'glyphs' object using util.inspect() just before it does `include glyphs.(id)`

Gist is here: https://gist.github.com/bitserf/dcf129ec74a430ba51e3

I can only find entry for `D_Zcaron` in the object, and not `Zcaron` (value of `id`). 

If `glyphs.(id)` performs a key-based lookup or a function call of some kind - I don't know what the syntax does :)

**Comment by be5invis** [comment]
Is it a platform-specific behavior about String.prototype.normalize? What's the result of `'Ž'.normalize('NFD')`? The `glyphs` hash does not contain any precomposed glyphs at all.

**Comment by be5invis** [comment]
Okay, I pushed a newer version with `String.prototype.normalize` replaced with `unorm`. Is this version working?
(You may install the new dependency `unorm` via `npm install` or `npm install unorm`)

**Comment by bitserf** [comment]
I have updated to latest code, installed `unorm` and ran build again. It is fixed, I have successfully built all fonts.

Here is the output of normalise in my REPL:

```shell
> '\u017D'.normalize('NFD').split('').map(function(c){ return c.charCodeAt(0).toString(16) })
[ '17d' ]
```

So it seems the default node.js package is the cause. By default, Homebrew installs a version of node.js where the normalize() function is just an identity function and returns its input, that is stupid in a global world.

If I re-install node.js, and recompile it with ICU support:

```shell
$ brew install node --with-icu4c
$ node
> > '\u017D'.normalize('NFD').split('').map(function(c){ return c.charCodeAt(0).toString(16) })
[ '5a', '30c' ]
```

Then I get the results you are expecting in the REPL, but I got some warnings at build time about libstdc++ version issues, probably because I am running Xcode 7.0, and OS X 10.11 (not released yet).

I think using `unorm` like you did is a better fix for cross platform, appears to work even if they have a bad node.js build.

Thanks!!!
